snuba/migrations/snuba_migrations/metrics/0001_metrics_sets.py

```diff
@@ -0,0 +1,160 @@
+from typing import Sequence
+
+from snuba.clickhouse.columns import (
+    AggregateFunction,
+    Array,
+    Column,
+    DateTime,
+    Nested,
+    UInt,
+)
+from snuba.clusters.storage_sets import StorageSetKey
+from snuba.datasets.storages.tags_hash_map import INT_TAGS_HASH_MAP_COLUMN
+from snuba.migrations import migration, operations, table_engines
+from snuba.migrations.columns import MigrationModifiers as Modifiers
+
+sets_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("metric_id", UInt(64)),
+    Column("granularity", UInt(32)),
+    Column("timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+]
+
+sets_mv_columns: Sequence[Column[Modifiers]] = [
+    Column("org_id", UInt(64)),
+    Column("project_id", UInt(64)),
+    Column("metric_id", UInt(64)),
+    Column("granularity", UInt(32)),
+    Column("timestamp", DateTime()),
+    Column("retention_days", UInt(16)),
+    Column(
+        "tags",
+        Nested([Column("key", Array(UInt(64))), Column("value", Array(UInt(64)))]),
+    ),
+    Column("value", AggregateFunction("uniqCombined64", [UInt(64)])),
+]
+
+
+matview_statement = """
+SELECT
+    org_id,
+    project_id,
+    metric_id,
+    60 as granularity,
+    toStartOfMinute(timestamp) as timestamp,
+    retention_days,
+    tags.key,
```

Comment:
This might be an offline discussion/documented somewhere I haven't seen. I'm confused about how the tag keys/values are going to be used. Wouldn't it make more sense to arrayJoin on them so the HLL is for a specific tag?

Reply:
If I get the question correctly, you are suggesting to arrayJoin on tags and have buckets organized this way:
```
original buckets
tags {A:b B:d}, values
tags {A:c}, values

after arrayJoin
tag: A value: b HLL
tag: A value: c HLL
tag: B value: d HLL
```

Unfortunately this won't work as we would not be able to answer the query: `SELECT uniq(users) where tags[A] = b and tags[b] = d`. Doing the arrayjoin on the tags we would loose the fact that tags A and B where on the same bucket.
So each combination of tags we receive from the user (its full set of dimension) needs to be stored as a row in the materialized view.